Fix number_to_words for floats that stringify in scientific notation

BBC-Esq opened the pull request:
## Summary

Fixes #226.

When ``str(num)`` produces a string like ``'1e-06'`` (e.g. for ``0.000001``, or for floats with magnitude ``>= 1e16``), the digits of the mantissa and exponent were being parsed as the number, yielding nonsense such as ``'one hundred and six'`` for ``0.000001``.

This routes ``float`` inputs through ``Decimal(repr(num))`` and formats with ``'f'`` to obtain the exact fixed-point representation before further processing. ``repr`` already yields the shortest round-trip form Python uses, so values whose ``str`` is not in scientific notation (``0.1``, ``999.3``, ``1.5``, …) round-trip without introducing binary-float artifacts. Only the scientific-notation cases get rewritten.

The conversion lives in a small helper, ``engine._normalize_number``, with the rationale documented in its docstring (per @jaraco's review).

## Why not the prior approach

An earlier revision of this PR formatted floats with ``f"{num:.17g}"`` / ``f"{num:.17f}"``. That fixed the reported case but exposed binary-float artifacts on common inputs:

- ``0.1`` → ``'0.10000000000000001'``
- ``999.3`` → ``'999.29999999999995'`` (broke ``test_lines``)
- ``0.3`` → 17-digit garbage ending in ``'...nine nine nine nine'``

``Decimal(repr(num))`` avoids the artifacts because ``repr(0.1) == '0.1'`` exactly, while still expanding ``repr(0.000001) == '1e-06'`` to ``'0.000001'``. Same approach as #236 (closed by its author without review).

## Behavior matrix

| input | output of ``_normalize_number`` |
|---|---|
| ``0.000001`` | ``'0.000001'`` (issue fixed) |
| ``-0.000001`` | ``'-0.000001'`` |
| ``0.1`` / ``0.2`` / ``0.3`` | ``'0.1'`` / ``'0.2'`` / ``'0.3'`` (no regression) |
| ``999.3`` / ``1000.3`` | ``'999.3'`` / ``'1000.3'`` (existing tests pass) |
| ``1e20`` | ``'100000000000000000000'`` |
| ``1e-20`` | ``'0.00000000000000000001'`` (full precision) |
| ``1.0`` | ``'1.0'`` (matches ``str(1.0)``) |
| ``int`` / ``Decimal`` / ``Fraction`` / ``str`` | ``str(num)`` (unchanged path) |

## Test plan

- [x] New regression test ``tests/test_numwords.py::test_issue_226_float_scientific_notation`` covers ``0.000001``, ``-0.000001``, parity with the string form, two common floats that must remain unaffected (``0.1``, ``999.3``), and a large-magnitude (``1e20``) case. Fails on ``main``, passes after this change.
- [x] Full ``pytest tests/`` passes (208 passed, 16 xfailed, 0 failed) — including ``test_lines`` which exercises ``999.3``-style floats through the threshold path.
- [x] ``ruff check`` introduces no new errors vs ``main``.
- [x] News fragment added at ``newsfragments/226.bugfix.rst``.

jaraco commented on inflect/__init__.py:
```diff
@@ -3839,23 +3839,28 @@ def number_to_words(  # noqa: C901
         self._number_args = {"andword": andword, "zero": zero, "one": one}
-        num = str(num)
+
+        # Handle scientific notation conversion
+        if isinstance(num, float):
+            formatted = f"{num:.17g}"
+            if 'e' in formatted.lower():
+                num = f"{num:.17f}".rstrip('0').rstrip('.')
+            else:
+                num = formatted
+        else:
+            num = str(num)
```
I'd recommend to extract this logic into a helper function/method, something like `num = _normalize_number(num)`. That way, the comment can be in the docstring. This approach also limits the cyclomatic complexity.

jaraco reviewed (changes requested):
Looking good! Can you add a test capturing the new expectation?

BBC-Esq commented:
Unfortunately due to my novice programming experience I don't know what "capturing the new expectation" means...Not joking sadly.

jaraco commented:
> Unfortunately due to my novice programming experience I don't know what "capturing the new expectation" means...Not joking sadly.

Understood. What I mean is that this change introduces a "new expectation" that inflect now handles numbers in scientific notation. I want to add a test that will fail on the current code in main but pass after applying the proposed changes. The tests in this project can exist in a number of places (probably in this order of preference where appropriate):

- static expected inflections in tests/inflections.txt (these are validated)
- unit tests in tests/test_*
- doctests in some functions or methods

To get started, you'll want to first validate that you can run the tests on `main` and get a passing result. See the [skeleton docs](https://blog.jaraco.com/skeleton/#running-tests) for guidance on running tests, and don't hesitate to let me know if you have questions.

You're also welcome to join my [Discord server](https://discord.gg/BJYAcfBp) for interactive support if that might help.

BBC-Esq commented:
Are you able to finish this pull request by chance?  Unfortunately I can't even though I appreciate you assigning the issue to me.

BBC-Esq committed:
Fix number_to_words for floats that stringify in scientific notation

When ``str(num)`` produces a string like ``'1e-06'`` (e.g. for
``0.000001`` or for floats with magnitude ``>= 1e16``), the digits of
the mantissa and exponent were being parsed as the number, yielding
nonsense such as ``'one hundred and six'`` for ``0.000001``.

Extract a ``_normalize_number`` helper that routes ``float`` inputs
through ``Decimal(repr(num))`` and formats with ``'f'`` to obtain the
exact fixed-point representation before further processing. ``repr``
already yields the shortest round-trip form, so values whose ``str``
is not in scientific notation (``0.1``, ``999.3``, ``1.5``, ...) are
unaffected — no binary-float artifacts are introduced.

Fixes #226

BBC-Esq commented:
Friendly ping, @jaraco — when you have a moment, could you take another look at this one? The rewrite addresses each of the points from the prior review (helper extraction, regression test, news fragment) and tests/lint are green. Happy to make any further adjustments if you'd like a different shape — otherwise hoping to get it across the line. Thanks!

BBC-Esq commented:
Closing this for now. I've merged the work into the ``main`` branch of my fork at https://github.com/BBC-Esq/inflect — anyone who wants the fix is welcome to pick it up from there at any time.

Thanks for the earlier feedback that shaped the current approach. If you'd like to revive this here later, the diff is small and self-contained: ``inflect/__init__.py`` adds ``from decimal import Decimal`` plus an ``engine._normalize_number`` staticmethod, with a regression test in ``tests/test_numwords.py`` and a news fragment at ``newsfragments/226.bugfix.rst``.
